### DISCUSSION TITLE
Policy behaving diffrently in Envoy proxy

### DISCUSSION THREAD

**Discussion by 3goats** [open-policy-agent/community#493] [Community]
So I have this policy which basically checks that the SAN UPN value is submitted CSR matches a claim within the JWT used for authorisation. 

https://play.openpolicyagent.org/p/1tmmorwMuD

The demo's input provide an `authorization1` and `authorization2` field. If you set the `http.headers.authorizationX`  in the token section to `authorization1` the fields match and the policy is allowed and it returns a 200 status code. 

If set to `authorization1` the fields don't match and the policy is not allowed and it returns a 407 status code.

```
token := {"valid": valid, "payload": payload} if {
    [_, encoded] := split(http.headers.authorization1, " ")
    [valid, _, payload] := io.jwt.decode_verify(encoded, {"secret": "secret"})
}
```

This seems to work as expected in the playground, however if I then move it to the Envoy proxy it returns a 401 instead of the 407 upon failure.

**Comment by anderseknert** [comment]
See the configuration options for the plugin here: https://www.openpolicyagent.org/docs/edge/envoy-introduction/#configuration

The `plugins["envoy_ext_authz_grpc"].path` is what you'll want to set. Given your playground policy, you'd want to set the path to `envoy/authz/result`, but note that it says:

> If the policy decision is an object, it **must** contain the allowed key set to either true or false to indicate if the request is allowed or not respectively.

So you'll want to add:

```rego
result["allowed"] := allow
```

Not sure if it solves the problem, but it was the one outstanding issue I noticed :)

**Comment by ashutosh-narkar** [comment]
Also you can define the HTTP response status code sent to the downstream client when a request is denied. Be default, Envoy returns a 403. More info [here](https://www.openpolicyagent.org/docs/edge/envoy-primer/#output-document).

**Reply by anderseknert** [reply]
That's already in his example policy. Or am I missing something?

**Reply by ashutosh-narkar** [reply]
My bad. Didn't look at the PG link before.

**Comment by 3goats** [comment]
Thanks I'll take a look at this tomorrow.

**Comment by 3goats** [comment]
OK so I set `--set=plugins.envoy_ext_authz_grpc.path=envoy/authz/allow` to `--set=plugins.envoy_ext_authz_grpc.path=envoy/authz/result` in the `app.yaml` file in the standalone tutorial. However, Envoy still returns a 403 instead of a 407 as defined in the policy. 


Am I missing something ?

**Reply by anderseknert** [reply]
I'm not sure, but I have had issues with `--set` values needing to be quoted in the past... perhaps something to try?

You could always sprinkle some `print` statements in your policy to see if anything is evaluated, I guess.

**Reply by anderseknert** [reply]
@ashutosh-narkar might have some ideas. I've not used Envoy in a long time.

**Comment by 3goats** [comment]
My `--set` values are already quoted.

**Comment by ashutosh-narkar** [comment]
I would first unit test the policy to ensure you're seeing the correct response. If that looks ok, then turn on debug logging to ensure you see the correct OPA policy result. At this point we've verified that the response OPA sends to Envoy is what we expect. Then you can look into Envoy logs to see if you notice anything in there. 

You could also add a log [here](https://github.com/open-policy-agent/opa-envoy-plugin/blob/main/internal/internal.go#L475) to print the complete response OPA sends to Envoy.
